# Métodos de optimización en bio-reactores
Aplicación de método de optimización a la mejora de producción de L-(-)-carnitina a partir de crotonobetaína en un biorreactor en régimen continuo con reciclaje celular.

La carnitina es un compuesto de amplio rango de aplicaciones médicas. Gran parte de su producción se obtiene por síntesis química, pero esta tiene muchos inconvenientes. Por estas razones, y su incremento en demanda, se optan otro tipo de rutas para sintetizar de una mejor manera. La _E. Coli_ es una excelente productora del compuesto, así mismo varias enzimas también están implicadas en la biosíntesis de la L-carnitina. Por lo que el uso de biorreactores es recomendado para tener un sistema de bioproducción en régimen contínuo con reciclaje celular. 

A partir de un modelo previo, se obtuvo la descripción del sistema llamado S-system con las siguientes equivalencias:

$X_1:G$ Concentración glicerol  
$X_2:Cr$ Concentración de crotonobetaína  
$X_3:C$ Concentración de carnitina  
$X_4:X$ Biomasa  
$X_5:Q$ Velocidad de extracción del biorreactor   
$X_6:G_{in}$  Concentración de glicerol
$X_7:Cr_{in}$ Concentración de crotonobetaína  
$X_8:\mu_{max}$  
$X_9:KAI$ Actividad de la carnitina deshidratasa  

Con ecuaciones diferenciales, los procesos metabólicos se pueden expresar con el sistema S-system donde que consta de una parte positiva y otra negativa. En la parte positiva, se consideran todos los elementos que influyen en la producción de algún producto,y  en la negativa son todos los elementos que influyen en la transformación del mismo. En este caso específico el órden cinético y las constantes de velocidad se obtuvieron a partir de datos experimentales previos, por lo que el conjunto de ecuaciones ya desarrollado aplicado a un sistema metabólico queda:

$\frac{dX_1}{dt}=X_5X_6-(1.0213)X_1^{0.9162}X_4^{0.5675}X_5^{0.4324}X_8^{0.5675}$

$\frac{dX_2}{dt}=(0.2438)X_3^{0.3538}X_4^{0.9394} X_5^{0.0605} X_7^{0.0605} X_9^{0.9292}-(0.464)X_2^{0.1424}X_4^{0.9643}X_5^{0.0356}X_9^{0.9537}$

$\frac{dX_3}{dt}=(0.3844)X_2^{0.1106} X_4 X_9^{0.989}-(0.2058)X_3^{0.3926}X_4^{0.9742}X_5^{0.0257}X_9^{0.9636}$

$\frac{dX_4}{dt}=(0.0053) X_1^{0.8524} X_4 X_8 -(0.08)X_4$

En estado estacionario, la derivadas de las variables respecto al tiempo se hacen cero. Por lo que las ecuaciones resultantes se pueden tomar como restricciones del sistema en estado estacionario, es importante hacer el analisis dentro de este estado ya que sus características no varían con el tiempo, y así mismo es una prueba de calidad que exista estabilidad del estado estacionario:

$0=X_5X_6-1.0213 X_1^{0.9162}X_4^{0.5675}X_5^{0.4324}X_8^{0.5675}$  
$0=0.2438 X_3^{0.3538} X_4^{0.9394} X_5^{0.0605} X_7^{0.0605} X_9^{0.9292}-0.464X_2^{0.1424}X_4^{0.9643}X_5^{0.0356}X_9^{0.9537}$

$0=0.3844 X_2^{0.1106} X_4 X_9^{0.989}-0.2058 X_3^{0.3926}X_4^{0.9742}X_5^{0.0257}X_9^{0.9636}$

$0=0.0053 X_1^{0.8524} X_4 X_8 -0.08 X_4$

Despejando se puede obtener:


$1.0213=\frac {X_5X_6}{X_1^{0.9162}X_4^{0.5675}X_5^{0.4324}X_8^{0.5675}}$  

$\frac{0.2438}{0.464} =\frac{X_2^{0.1424}X_4^{0.9643}X_5^{0.0356}X_9^{0.9537}}{X_3^{0.3538} X_4^{0.9394} X_5^{0.0605} X_7^{0.0605} X_9^{0.9292}}$

$\frac{0.3844}{0.2058}= \frac{X_3^{0.3926}X_4^{0.9742}X_5^{0.0257}X_9^{0.9636}}{ X_2^{0.1106} X_4 X_9^{0.989}}$

$\frac{0.08}{0.0053}= X_1^{0.8524}X_8$


Al aplicar logaritmo natural a ambos lados de la ecuación, y reducir en términos semejantes, se puede desarrollar el siguiente sistema de ecuaciones.


$0.0211= -0.9162ln(X_1)-0.5675ln(X_4)+0.5675ln(X_5)+ln(X_6)-0.5675ln(X_8)$  

$0.6433 =-.1424 ln(X_2)+{0.3538}ln(X_3)-{0.0248}ln(X_4)+{0.0248}ln(X_5)+0.0605ln(x_7)-.0245ln(X_9)$


$0.6246=-.1106 ln(X_2)+{0.3926}ln(X_3)-{0.0257}ln(X_4)+{0.0257}ln(X_5)-.0254ln(X_9)$

$2.7143= 0.8524ln(X_1)+ln(X_8)$

Haciéndo un cambio de variable sencillo: $Y_i=ln(X_i)$, se puede llegar a una expresión más agradable, con una interpretación lineal:


$0.0211= -0.9162 Y_1-0.5675 Y_4+0.5675 Y_5+ Y_6-0.5675 Y_8$  

$0.6433 =-0.1424 Y_2+0.3538 Y_3-0.0248 Y_4+0.0248Y_5+0.0605Y_7-0.0245Y_9$

$0.6246=-0.1106 Y_2+0.3926 Y_3-0.0257 Y_4+0.0257 Y_5-0.0254 Y_9$

$2.7143= 0.8524 Y_1+ Y_8$

Se define las restricciones del valor que pueden obtener las variables y parámetros del modelo. En este caso se definió que el rango de variación será entre los 0.5 y 1.5 veces los valores basales, exceptuando el $X_4$ que se le permitió variar hasta 2 veces el valor basal, estos valores estaban dados ya en tablas:

Variable| Valor Basal| Valor Basal Factor Mínimo| Valor basal Factor Máximo|
----|-----
$X_1$|43.28|21.64|64.92
$X_2$|29.49|14.745|44.235
$X_3$|20.51|10.255|30.765
$X_4$|12.18|6.09|24.36
$X_5$|1|.5|1.5
$X_6$|100|50|150
$X_7$|50|25|75
$X_8$|.603|.3015|0.9045
$X_9$|123.3|61.65|184.95


Las restricciones se dieron ya en escala logarítmica

$$3.0764 \leq Y_1 \leq 4.1732$$  
$$2.6909 \leq Y_2 \leq 3.7895$$  
$$2.3277 \leq Y_3 \leq 3.4263$$  
$$1.8068 \leq Y_4 \leq 3.1931$$  
$$-0.6931 \leq Y_5 \leq 0.4054$$  
$$3.9120 \leq Y_6 \leq 5.0106$$  
$$3.2188 \leq Y_7 \leq 4.3174$$  
$$-1.1988 \leq Y_8 \leq -0.1002$$  
$$4.1214 \leq Y_9 \leq 6.4240$$  

Lo que se quiere optimizar en este caso es la velocidad en la cual la carnitina se genera en el biorreactor:

$$Maximizar:V_c=X_3X_5 \rightarrow Maximizar:Y_3+Y_5$$

Comentario: Esta función objetivo es equivalente ya que el logaritmo es creciente y no debería de cambiar el signo del conjunto. 

In [10]:
from pulp import *

#Se crea la variable de problema, y se indica si se quiere maximizar o minimixar:
prob=LpProblem("Maximizar la velocidad en la cual la carnitina se genera en el biorreactor", LpMaximize)

#Se definen las variables a ser utilizadas
Y1=LpVariable("Y1: Logaritmo natural de: Concentración de glicerol (G)",lowBound=0, upBound=4.1732)
Y2=LpVariable("Y2: Logaritmo natural de: Concentración de crotonobetaína (Cr)",lowBound=2.6909, upBound=3.7895)
Y3=LpVariable("Y3: Logaritmo natural de: Concentración de carnitina (C)",lowBound=2.3277, upBound=3.4263)
Y4=LpVariable("Y4: Logaritmo natural de: Biomasa (X)",lowBound=1.8068, upBound=3.1931)
Y5=LpVariable("Y5: Logaritmo natural de: Velocidad de extracción del biorreactor (Q) ",lowBound=-0.6931, upBound=0.4054)
Y6=LpVariable("Y6: Logaritmo natural de: Concentración de glicerol (G_in)",lowBound=3.9120, upBound=5.0106)
Y7=LpVariable("Y7: Logaritmo natural de: Concentración de crotonobetaína (C_in)",lowBound=3.2188, upBound=4.3174)
Y8=LpVariable("Y8: Logaritmo natural de: mu_max",lowBound=-1.1988, upBound= -0.1002)
Y9=LpVariable("Y9: Logaritmo natural de: Actividad de la carnitina deshidratasa (KAI)", lowBound=4.1214 , upBound=6.4240)

#Se crea la función objetivo:
Funcion_objetivo=Y3+Y5

#Se crean las restricciones
restriccion1=-0.9162*Y1-0.5675*Y4+0.5675*Y5+Y6-0.5675*Y8 == 0.0211
restriccion2=-0.1424*Y2+0.3538*Y3-0.0248*Y4+0.0248*Y5+0.0605*Y7-0.0245*Y9 == 0.6433
restriccion3=-0.1106*Y2+0.3926*Y3-0.0257*Y4+0.0257*Y5-0.0254*Y9 == 0.6246
restriccion4= 0.8524*Y1 + Y8 == 2.7143

#Agregar la función objetivo al problema:
prob += Funcion_objetivo

#Agregar las restricciones al problema:
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Valor Óptimo: Velocidad =", value(prob.objective))

Status: Optimal
Y1:_Logaritmo_natural_de:_Concentración_de_glicerol_(G) = 3.3018536
Y2:_Logaritmo_natural_de:_Concentración_de_crotonobetaína_(Cr) = 3.7895
Y3:_Logaritmo_natural_de:_Concentración_de_carnitina_(C) = 3.2565772
Y4:_Logaritmo_natural_de:_Biomasa_(X) = 3.1931
Y5:_Logaritmo_natural_de:_Velocidad_de_extracción_del_biorreactor_(Q)_ = 0.4054
Y6:_Logaritmo_natural_de:_Concentración_de_glicerol_(G_in) = 4.5714145
Y7:_Logaritmo_natural_de:_Concentración_de_crotonobetaína_(C_in) = 4.2524093
Y8:_Logaritmo_natural_de:_mu_max = -0.1002
Y9:_Logaritmo_natural_de:_Actividad_de_la_carnitina_deshidratasa_(KAI) = 6.424
Valor Óptimo: Velocidad = 3.6619772


Ya que la fórmula que se tenía para las variables era $Y_i=ln(X_i)$ Para regresar a obtener los valores X_i lo único que se tiene que hacer para obtener $X_i$ es: $e^{Y_i}=X_i$. 

Esto hace que los valores obtenidos cambien a una escala lineal, dando como resultado los siguientes datos:

$Y_i$| Valor |$X_i$| Valor| Nombre|Unidades
:-----|------
$Y_1$|3.3018|$X_1$|27.162|Concentración de glicerol ($G$)|$mM$
$Y_2$|3.7895|$X_2$|44.234|Concentración de crotonobetaína ($Cr$)|$mM$
$Y_3$|3.2565|$X_3$|25.959| Concentración de carnitina ($C$)|$mM$
$Y_4$|3.1931|$X_4$|24.364|Biomasa ($X$)|$g/L$
$Y_5$|0.4054|$X_5$|1.4999|Velocidad de extracción del biorreactor ($Q$)|$L/h^{-1}$
$Y_6$|4.5714|$X_6$|96.679|Concentración de glicerol ($G_{in}$)|$mM$
$Y_7$|4.2534|$X_7$|70.344|Concentración de crotonobetaína ($C_{in}$)|$mM$
$Y_8$|-0.1002|$X_8$|0.9746|$\mu_{max}$|$h^{-1}$
$Y_9$|6.4240|$X_9$|616.46|Actividad de la carnitina deshidratasa ($KAI$)|$UhL^{-1}$
    **Función Obj.**| 3.6619772|**Función Obj.**|38.93825553|Velocidad de Producción| $mM/h^{-1}$